File: core/object-model/js/src/manifest-stream.js
import { createHash } from 'node:crypto';
import { encodeCanonical } from './cbor.js';
import { fail, isOgvcsError } from './errors.js';
import { createObjectHashWriter } from './hash.js';
import { configuredHardLimit, enforceHardLimit, hardLimitMaximum } from './hard-limits.js';
import { profileDecision, registryAssignmentDecision } from './registry.js';
import { validateKnownSchema } from './schema.js';
import { Digest, KIND_NAMES, ObjectRef, ProfileRef, equalBytes } from './types.js';
import {
  ResourceGuard, asBytes, asCount, asLimit, cborHeader, checkedBigUint,
  exactMap, guardedAsyncIterable, toAsyncIterable, writeFully
} from './scale-util.js';

const MAX_MANIFEST_PARTS = hardLimitMaximum('manifest-chunks');
const MAX_CHUNK_BYTES = BigInt(hardLimitMaximum('chunk-payload-bytes'));
const MAX_LOGICAL_BYTES = BigInt(hardLimitMaximum('logical-file-bytes'));
const MAX_METADATA_BYTES = hardLimitMaximum('metadata-payload-bytes');
const MAX_GENERIC_VALUE_BYTES = hardLimitMaximum('generic-text-or-byte-value-bytes');
const CHUNKING_FAMILY = 'chunking';
const RECORD_OVERHEAD = 96;

function namesFor(registry) { return registry?.kindNames ?? KIND_NAMES; }

function digestValue(value) {
  if (value instanceof Digest) return value;
  return Digest.fromMap(value);
}

function chunkProfile(value, registry, operation) {
  const ref = value instanceof ProfileRef ? value : ProfileRef.fromMap(value);
  if (registry) {
    const decision = profileDecision(registry, ref, operation);
    if (decision.family !== CHUNKING_FAMILY) fail('SCHEMA_FIELD_INVALID', { layer: 2 });
  }
  return ref;
}

function encodeResident(value, options, remaining = options.maxMemoryBytes, overrides = {}) {
  const memoryBound = Math.min(remaining, options.maxBytes);
  try {
    return encodeCanonical(value, {
      ...overrides,
      maxBytes: memoryBound,
      maxWorkingBytes: Math.min(options.maxMemoryBytes, remaining)
    });
  } catch (error) {
    if (isOgvcsError(error, 'LIMIT_METADATA_BYTES') && memoryBound < options.maxBytes) {
      fail('LIMIT_MEMORY', { layer: 1, cause: error });
    }
    throw error;
  }
}

function optionsFor(input) {
  const maxItems = Math.min(
    configuredHardLimit('manifest-chunks', input.hardLimits?.['manifest-chunks']),
    configuredHardLimit('manifest-chunks', asLimit(input.maxItems, MAX_MANIFEST_PARTS))
  );
  enforceHardLimit(undefined, 'manifest-chunks', input.partCount, {
    maximum: maxItems, code: 'LIMIT_COUNT', layer: 2
  });
  const partCount = asCount(input.partCount, maxItems);
  const maxBytes = Math.min(
    configuredHardLimit('metadata-payload-bytes', input.hardLimits?.['metadata-payload-bytes']),
    configuredHardLimit('metadata-payload-bytes', asLimit(input.maxBytes, MAX_METADATA_BYTES))
  );
  const maxMemoryBytes = asLimit(input.maxMemoryBytes, 67_108_864);
  const guard = new ResourceGuard({ maxTimeMs: input.maxTimeMs, maxMemoryBytes });
  guard.time();
  if (!Array.isArray(input.requiredFeatures ?? [])) fail('SCHEMA_FIELD_INVALID', { layer: 2 });
  enforceHardLimit(undefined, 'logical-file-bytes', input.logicalLength, {
    maximum: input.hardLimits?.['logical-file-bytes'], code: 'LIMIT_LOGICAL_BYTES', layer: 2
  });
  const logicalLength = checkedBigUint(input.logicalLength, MAX_LOGICAL_BYTES, 'LIMIT_LOGICAL_BYTES');
  const profile = chunkProfile(input.chunkProfile, input.registry, input.operation ?? 'conformance');
  return {
    ...input, maxItems, partCount, maxBytes, maxMemoryBytes, guard, logicalLength, profile,
    requiredFeatures: input.requiredFeatures ?? [], operation: input.operation ?? 'conformance'
  };
}

function manifestPrefix(options, digest) {
  const digestMap = digest.toMap();
  const profileMap = options.profile.toMap();
  const pieces = [];
  let resident = 0;
  const push = part => {
    options.guard.memory(resident + part.length);
    resident += part.length;
    pieces.push(part);
  };
  const pushValue = (value, overrides) => push(encodeResident(
    value, options, options.maxMemoryBytes - resident, overrides
  ));
  push(cborHeader(5, options.extensions === undefined ? 7 : 8));
  pushValue(0); pushValue(1); pushValue(1); pushValue(2); pushValue(2); pushValue(options.requiredFeatures);
  if (options.extensions !== undefined) { pushValue(3); pushValue(options.extensions); }
  pushValue(16); pushValue(options.logicalLength); pushValue(17); pushValue(digestMap);
  pushValue(18); pushValue(profileMap); pushValue(19); push(cborHeader(4, options.partCount));

  const common = new Map([[0, 1], [1, 2], [2, options.requiredFeatures], [16, options.logicalLength],
    [17, digestMap], [18, profileMap], [19, []]]);
  if (options.extensions !== undefined) common.set(3, options.extensions);
  // Validate the common fields and profile family with an empty, internally
  // consistent manifest. The actual streamed count and sum are checked below.
  const empty = new Map(common);
  empty.set(16, 0);
  empty.set(17, new Digest(1, new Uint8Array(createHash('sha256').digest())).toMap());
  validateKnownSchema(empty, 2, {
    registry: options.registry, operation: options.operation, hardLimits: options.hardLimits
  });
  return pieces;
}

function validatePart(value, options) {
  exactMap(value, [0, 1]);
  const reference = ObjectRef.fromMap(value.get(0), namesFor(options.registry));
  if (reference.kind !== 1) {
    fail('OBJECT_REFERENCE_KIND_MISMATCH', { layer: 2, stage: 'known-schema' });
  }
  if (options.registry) {
    registryAssignmentDecision(options.registry, 'object-kinds', 1, options.operation);
    registryAssignmentDecision(options.registry, 'hash-algorithms', 1, options.operation);
  }
  enforceHardLimit(undefined, 'chunk-payload-bytes', value.get(1), {
    maximum: options.hardLimits?.['chunk-payload-bytes'], code: 'MANIFEST_CHUNK_LENGTH_INVALID', layer: 2
  });
  const length = checkedBigUint(value.get(1), MAX_CHUNK_BYTES, 'MANIFEST_CHUNK_LENGTH_INVALID', 1n);
  const encoded = encodeResident(value, options, options.maxMemoryBytes, {
    maxValueBytes: MAX_GENERIC_VALUE_BYTES,
    maxContainerItems: 8
  });
  options.guard.memory(encoded.length + RECORD_OVERHEAD);
  return { reference, length, encoded };
}

async function iterableFromFactory(parts, pass, guard) {
  if (typeof parts === 'function') {
    return toAsyncIterable(await guard.wait(signal => parts(pass, { signal })));
  }
  if (pass !== 1) fail('SCHEMA_FIELD_INVALID', { layer: 2 });
  return toAsyncIterable(parts);
}

async function providerBytes(provider, reference, context, guard) {
  if (typeof provider !== 'function') fail('SCHEMA_FIELD_INVALID', { layer: 2 });
  const supplied = await guard.wait(signal => provider(reference, Object.freeze({ ...context, signal })));
  if (supplied instanceof Uint8Array || ArrayBuffer.isView(supplied) || supplied instanceof ArrayBuffer) {
    return [asBytes(supplied)];
  }
  return toAsyncIterable(supplied);
}

class ChunkVerifier {
  constructor(options, whole) {
    this.options = options;
    this.whole = whole;
    this.cache = new Map();
    this.cachedBytes = 0;
    this.contentBytes = 0n;
    this.providerReads = 0;
    this.maxCacheBytes = Math.min(asLimit(options.maxChunkCacheBytes, Math.floor(options.maxMemoryBytes / 2)),
      options.maxMemoryBytes);
  }

  async consume(part, index) {
    const key = part.reference.toString();
    const cached = this.cache.get(key);
    if (cached) {
      if (BigInt(cached.length) !== part.length) fail('MANIFEST_CHUNK_LENGTH_INVALID', { layer: 3 });
      this.whole.update(cached);
      this.contentBytes += BigInt(cached.length);
      return;
    }
    const projectedCache = this.cachedBytes + (this.cache.size + 1) * RECORD_OVERHEAD + Number(part.length);
    const shouldCache = projectedCache <= this.maxCacheBytes;
    if (shouldCache) this.options.guard.memory(projectedCache + part.encoded.length + RECORD_OVERHEAD);
    const retained = shouldCache ? new Uint8Array(Number(part.length)) : undefined;
    const chunkHash = createObjectHashWriter(1, {
      maxChunkBytes: Number(part.length),
      registry: namesFor(this.options.registry)
    });
    let length = 0n;
    this.providerReads += 1;
    const source = await providerBytes(this.options.chunkProvider, part.reference,
      { index, declaredLength: part.length }, this.options.guard);
    for await (const raw of guardedAsyncIterable(source, this.options.guard)) {
      this.options.guard.time();
      const bytes = asBytes(raw, 3);
      if (BigInt(bytes.length) > part.length - length) fail('MANIFEST_CHUNK_LENGTH_INVALID', { layer: 3 });
      if (retained) retained.set(bytes, Number(length));
      length += BigInt(bytes.length);
      this.contentBytes += BigInt(bytes.length);
      chunkHash.update(bytes);
      this.whole.update(bytes);
    }
    if (length !== part.length) fail('MANIFEST_CHUNK_LENGTH_INVALID', { layer: 3 });
    const actual = chunkHash.finish();
    if (!equalBytes(actual.digest, part.reference.digest)) fail('OBJECT_ID_MISMATCH', { layer: 1 });
    if (retained) {
      this.cachedBytes += retained.length;
      this.options.guard.memory(this.cachedBytes + (this.cache.size + 1) * RECORD_OVERHEAD +
        part.encoded.length + RECORD_OVERHEAD);
      this.cache.set(key, retained);
    }
  }
}

async function scanParts(options, { pass, verifyContent, consume }) {
  let count = 0;
  let logical = 0n;
  const transcript = createHash('sha256');
  const whole = verifyContent ? createHash('sha256') : undefined;
  const verifier = verifyContent ? new ChunkVerifier(options, whole) : undefined;
  const source = await iterableFromFactory(options.parts, pass, options.guard);
  for await (const raw of guardedAsyncIterable(source, options.guard)) {
    options.guard.time();
    count += 1;
    if (count > options.maxItems) {
      fail('LIMIT_COUNT', { layer: 1, stage: 'configured-resource-preflight' });
    }
    if (count > options.partCount) fail('SCHEMA_FIELD_INVALID', { layer: 2 });
    const part = validatePart(raw, options);
    logical += part.length;
    enforceHardLimit(undefined, 'logical-file-bytes', logical, {
      maximum: options.hardLimits?.['logical-file-bytes'], code: 'LIMIT_LOGICAL_BYTES', layer: 2
    });
    transcript.update(part.encoded);
    if (verifier) await verifier.consume(part, count - 1);
    if (consume) await consume(part.encoded);
  }
  if (count !== options.partCount) fail('SCHEMA_FIELD_INVALID', { layer: 2 });
  if (logical !== options.logicalLength || (options.logicalLength === 0n && count !== 0)) {
    fail('MANIFEST_LENGTH_MISMATCH', { layer: 2 });
  }
  return {
    count,
    logical,
    transcript: new Uint8Array(transcript.digest()),
    contentDigest: whole ? new Uint8Array(whole.digest()) : undefined,
    contentBytes: verifier?.contentBytes ?? 0n,
    providerReads: verifier?.providerReads ?? 0,
    cachedChunks: verifier?.cache.size ?? 0,
    cachedBytes: verifier?.cachedBytes ?? 0
  };
}

function createEmitter(options) {
  const hash = createObjectHashWriter(2, { maxMetadataBytes: options.maxBytes, registry: namesFor(options.registry) });
  let bytes = 0;
  return {
    async emit(part) {
      options.guard.time();
      if (part.length > options.maxBytes - bytes) fail('LIMIT_METADATA_BYTES', { layer: 1 });
      bytes += part.length;
      hash.update(part);
      await writeFully(options.sink, part, { guard: options.guard });
    },
    finish() { return { reference: hash.finish(), bytes }; }
  };
}

function stableSummary(reference, options, bytes) {
  return Object.freeze({
    format: 1,
    kind: 2,
    objectRef: reference.toString(),
    partCount: options.partCount,
    logicalBytes: options.logicalLength.toString(),
    metadataBytes: bytes
  });
}

/**
 * Validate and write ContentManifestV1 without retaining its part array.
 *
 * With a supplied wholeFileDigest, `parts` may be any iterable. Providing a
 * chunkProvider verifies every referenced byte and the whole-file digest.
 * Without a supplied digest, `parts` must be a repeatable factory: pass 1
 * validates content and derives the digest; pass 2 replays metadata, checks a
 * SHA-256 structural transcript, and writes the canonical object. This hashes
 * large logical content only once.
 */
export async function writeContentManifest(input) {
  const options = optionsFor(input ?? {});
  const derive = options.wholeFileDigest === undefined;
  const verifyContent = derive || options.verifyContent === true || options.chunkProvider !== undefined ||
    options.partCount === 0;
  if (derive && typeof options.parts !== 'function') fail('SCHEMA_FIELD_INVALID', { layer: 2 });
  if (verifyContent && options.partCount > 0 && typeof options.chunkProvider !== 'function') {
    fail('SCHEMA_FIELD_INVALID', { layer: 2 });
  }

  let digest;
  let verification;
  if (derive) {
    verification = await scanParts(options, { pass: 1, verifyContent: true });
    digest = new Digest(1, verification.contentDigest);
  } else {
    digest = digestValue(options.wholeFileDigest);
  }

  const prefix = manifestPrefix(options, digest);
  options.guard.memory(prefix.reduce((total, part) => total + part.length, 0));
  const emitter = createEmitter(options);
  for (const part of prefix) await emitter.emit(part);
  const outputPass = derive ? 2 : 1;
  const scanned = await scanParts(options, {
    pass: outputPass,
    verifyContent: !derive && verifyContent,
    consume: part => emitter.emit(part)
  });

  if (derive && !equalBytes(verification.transcript, scanned.transcript)) {
    fail('SCHEMA_FIELD_INVALID', { layer: 2 });
  }
  if (!derive && verifyContent && !equalBytes(scanned.contentDigest, digest.bytes)) {
    fail('MANIFEST_FILE_DIGEST_MISMATCH', { layer: 3 });
  }
  const effectiveVerification = derive ? verification : scanned;
  const finished = emitter.finish();
  return Object.freeze({
    objectRef: finished.reference,
    wholeFileDigest: digest,
    summary: stableSummary(finished.reference, options, finished.bytes),
    verification: Object.freeze({
      contentVerified: verifyContent,
      contentBytesRead: effectiveVerification.contentBytes.toString(),
      providerReads: effectiveVerification.providerReads,
      cachedChunks: effectiveVerification.cachedChunks,
      cachedBytes: effectiveVerification.cachedBytes
    }),
    metrics: Object.freeze({ elapsedMilliseconds: options.guard.elapsedMilliseconds() })
  });
}

export const MANIFEST_STREAM_LIMITS = Object.freeze({
  maxParts: MAX_MANIFEST_PARTS,
  maxChunkBytes: MAX_CHUNK_BYTES,
  maxLogicalBytes: MAX_LOGICAL_BYTES,
  maxMetadataBytes: MAX_METADATA_BYTES
});

File: core/object-model/rust/src/manifest_stream.rs
use std::{
    io::Write,
    time::{Duration, Instant},
};

use crate::{
    hard_limits::{
        enforce_hard_limit_context, MAX_CHUNK_BYTES, MAX_LOGICAL_FILE_BYTES, MAX_MANIFEST_CHUNKS,
        MAX_METADATA_BYTES,
    },
    Error, ErrorCode, ObjectHashWriter, ObjectKind, ObjectRef, Operation, ProfileRef, Registry,
    RegistryAssignment, Result, Sha256Writer, ValidationStage,
};

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct ManifestStreamPart {
    pub chunk: ObjectRef,
    pub length: u64,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct ManifestStreamLimits {
    pub max_parts: u64,
    pub max_output_bytes: u64,
    pub max_logical_bytes: u64,
    pub max_chunk_bytes: u64,
    /// Largest single source slice accepted by the verifier. The verifier
    /// never retains source slices or complete chunk payloads.
    pub max_memory_bytes: usize,
    pub max_elapsed: Option<Duration>,
}

impl Default for ManifestStreamLimits {
    fn default() -> Self {
        Self {
            max_parts: MAX_MANIFEST_CHUNKS,
            max_output_bytes: MAX_METADATA_BYTES,
            max_logical_bytes: MAX_LOGICAL_FILE_BYTES,
            max_chunk_bytes: MAX_CHUNK_BYTES,
            max_memory_bytes: 8 * 1024 * 1024,
            max_elapsed: None,
        }
    }
}

impl ManifestStreamLimits {
    fn constrained(self) -> Self {
        Self {
            max_parts: self.max_parts.min(MAX_MANIFEST_CHUNKS),
            max_output_bytes: self.max_output_bytes.min(MAX_METADATA_BYTES),
            max_logical_bytes: self.max_logical_bytes.min(MAX_LOGICAL_FILE_BYTES),
            max_chunk_bytes: self.max_chunk_bytes.min(MAX_CHUNK_BYTES),
            ..self
        }
    }
}

/// Contains only counts, lengths, and durable digests. Source bytes are never
/// retained or included in diagnostics.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct ManifestStreamSummary {
    pub object_ref: ObjectRef,
    pub parts: u64,
    pub payload_bytes: u64,
    pub logical_bytes: u64,
    pub whole_file_digest: [u8; 32],
}

/// Supplies a chunk as one or more borrowed slices. Implementations may reuse
/// one fixed buffer for every part; the verifier consumes each slice before
/// `stream_chunk` continues.
pub trait ManifestChunkSource {
    fn stream_chunk(
        &mut self,
        part_index: u64,
        part: &ManifestStreamPart,
        consume: &mut dyn FnMut(&[u8]) -> Result<()>,
    ) -> Result<()>;
}

impl<F> ManifestChunkSource for F
where
    F: FnMut(u64, &ManifestStreamPart, &mut dyn FnMut(&[u8]) -> Result<()>) -> Result<()>,
{
    fn stream_chunk(
        &mut self,
        part_index: u64,
        part: &ManifestStreamPart,
        consume: &mut dyn FnMut(&[u8]) -> Result<()>,
    ) -> Result<()> {
        self(part_index, part, consume)
    }
}

struct Budget {
    limits: ManifestStreamLimits,
    started: Instant,
}

impl Budget {
    fn new(limits: ManifestStreamLimits) -> Result<Self> {
        let result = Self {
            limits: limits.constrained(),
            started: Instant::now(),
        };
        result.check_time()?;
        Ok(result)
    }

    fn check_time(&self) -> Result<()> {
        if self
            .limits
            .max_elapsed
            .is_some_and(|maximum| self.started.elapsed() >= maximum)
        {
            return Err(Error::new(ErrorCode::LimitTime));
        }
        Ok(())
    }

    fn check_count(&self, count: u64) -> Result<()> {
        manifest_limit(
            "manifest-chunks",
            count,
            self.limits.max_parts,
            ErrorCode::LimitCount,
            2,
        )
    }
}

struct CanonicalSink<W> {
    writer: W,
    hash: ObjectHashWriter,
    written: u64,
    maximum: u64,
}

impl<W: Write> CanonicalSink<W> {
    fn new(writer: W, maximum: u64) -> Self {
        Self {
            writer,
            hash: ObjectHashWriter::new(
                ObjectKind::ContentManifest,
                MAX_CHUNK_BYTES as usize,
                usize::try_from(maximum).unwrap_or(usize::MAX),
            ),
            written: 0,
            maximum,
        }
    }

    fn bytes(&mut self, bytes: &[u8]) -> Result<()> {
        let length =
            u64::try_from(bytes.len()).map_err(|_| Error::new(ErrorCode::LimitMetadataBytes))?;
        manifest_limit(
            "metadata-payload-bytes",
            self.written.saturating_add(length),
            self.maximum,
            ErrorCode::LimitMetadataBytes,
            1,
        )?;
        self.writer
            .write_all(bytes)
            .map_err(|_| Error::new(ErrorCode::SchemaFieldInvalid))?;
        self.hash.update(bytes)?;
        self.written += length;
        Ok(())
    }

    fn head(&mut self, major: u8, value: u64) -> Result<()> {
        let (buffer, length) = cbor_head(major, value);
        self.bytes(&buffer[..length])
    }

    fn finish(self) -> Result<(ObjectRef, u64)> {
        Ok((self.hash.finish()?, self.written))
    }
}

struct ContentPass {
    logical_bytes: u64,
    whole_file_digest: [u8; 32],
    part_transcript: [u8; 32],
}

struct PartPass {
    logical_bytes: u64,
    part_transcript: [u8; 32],
}

/// Preflights the lightweight part iterator, streams every referenced byte
/// once, derives the whole-file digest, then replays the iterator to emit a
/// canonical `ContentManifestV1`. The factory must return the same ordered
/// parts on all three calls.
#[allow(clippy::too_many_arguments)]
pub fn encode_content_manifest_stream<W, F, I, S>(
    writer: W,
    declared_parts: u64,
    parts_factory: F,
    chunk_profile: &ProfileRef,
    source: &mut S,
    registry: &Registry,
    operation: Operation,
    limits: ManifestStreamLimits,
) -> Result<ManifestStreamSummary>
where
    W: Write,
    F: FnMut() -> I,
    I: IntoIterator<Item = ManifestStreamPart>,
    S: ManifestChunkSource + ?Sized,
{
    encode_manifest_internal(
        writer,
        declared_parts,
        parts_factory,
        chunk_profile,
        source,
        registry,
        operation,
        limits,
        None,
    )
}

/// As [`encode_content_manifest_stream`], additionally requiring an exact
/// caller-declared logical length and whole-file SHA-256 digest.
#[allow(clippy::too_many_arguments)]
pub fn encode_and_verify_content_manifest_stream<W, F, I, S>(
    writer: W,
    declared_parts: u64,
    parts_factory: F,
    chunk_profile: &ProfileRef,
    expected_logical_bytes: u64,
    expected_whole_file_digest: [u8; 32],
    source: &mut S,
    registry: &Registry,
    operation: Operation,
    limits: ManifestStreamLimits,
) -> Result<ManifestStreamSummary>
where
    W: Write,
    F: FnMut() -> I,
    I: IntoIterator<Item = ManifestStreamPart>,
    S: ManifestChunkSource + ?Sized,
{
    encode_manifest_internal(
        writer,
        declared_parts,
        parts_factory,
        chunk_profile,
        source,
        registry,
        operation,
        limits,
        Some((expected_logical_bytes, expected_whole_file_digest)),
    )
}

#[allow(clippy::too_many_arguments)]
fn encode_manifest_internal<W, F, I, S>(
    writer: W,
    declared_parts: u64,
    mut parts_factory: F,
    chunk_profile: &ProfileRef,
    source: &mut S,
    registry: &Registry,
    operation: Operation,
    limits: ManifestStreamLimits,
    expected: Option<(u64, [u8; 32])>,
) -> Result<ManifestStreamSummary>
where
    W: Write,
    F: FnMut() -> I,
    I: IntoIterator<Item = ManifestStreamPart>,
    S: ManifestChunkSource + ?Sized,
{
    let budget = Budget::new(limits)?;
    budget.check_count(declared_parts)?;
    validate_manifest_assignments(registry, operation)?;
    validate_chunk_profile(chunk_profile, registry, operation)?;
    if let Some((logical, _)) = expected {
        manifest_limit(
            "logical-file-bytes",
            logical,
            budget.limits.max_logical_bytes,
            ErrorCode::LimitLogicalBytes,
            2,
        )?;
    }

    let preflight = preflight_parts(declared_parts, parts_factory(), &budget)?;
    if expected.is_some_and(|(expected_length, _)| expected_length != preflight.logical_bytes) {
        return Err(Error::new(ErrorCode::ManifestLengthMismatch));
    }
    let pass = verify_content(declared_parts, parts_factory(), source, &budget)?;
    if pass.logical_bytes != preflight.logical_bytes
        || pass.part_transcript != preflight.part_transcript
    {
        return Err(Error::new(ErrorCode::ManifestLengthMismatch));
    }
    if let Some((expected_length, expected_digest)) = expected {
        if pass.logical_bytes != expected_length {
            return Err(Error::new(ErrorCode::ManifestLengthMismatch));
        }
        if pass.whole_file_digest != expected_digest {
            return Err(Error::new(ErrorCode::ManifestFileDigestMismatch));
        }
    }

    budget.check_time()?;
    let mut sink = CanonicalSink::new(writer, budget.limits.max_output_bytes);
    sink.head(5, 7)?;
    sink.head(0, 0)?;
    sink.head(0, 1)?;
    sink.head(0, 1)?;
    sink.head(0, 2)?;
    sink.head(0, 2)?;
    sink.head(4, 0)?;
    sink.head(0, 16)?;
    sink.head(0, pass.logical_bytes)?;
    sink.head(0, 17)?;
    write_typed_digest(&mut sink, &pass.whole_file_digest)?;
    sink.head(0, 18)?;
    write_profile(&mut sink, chunk_profile)?;
    sink.head(0, 19)?;
    sink.head(4, declared_parts)?;

    let mut second_count = 0u64;
    let mut second_logical = 0u64;
    let mut second_transcript = Sha256Writer::new();
    for part in parts_factory() {
        budget.check_time()?;
        second_count = second_count
            .checked_add(1)
            .ok_or_else(|| Error::new(ErrorCode::LimitCount))?;
        if second_count > declared_parts {
            return Err(Error::new(ErrorCode::LimitCount));
        }
        validate_part(&part, &budget)?;
        second_logical = add_logical(second_logical, part.length, &budget)?;
        let encoded = encode_part(&part);
        second_transcript.update(&encoded);
        sink.bytes(&encoded)?;
    }
    if second_count != declared_parts {
        return Err(Error::new(ErrorCode::LimitCount));
    }
    if second_logical != pass.logical_bytes || second_transcript.finish() != pass.part_transcript {
        return Err(Error::new(ErrorCode::ManifestLengthMismatch));
    }

    let (object_ref, payload_bytes) = sink.finish()?;
    Ok(ManifestStreamSummary {
        object_ref,
        parts: declared_parts,
        payload_bytes,
        logical_bytes: pass.logical_bytes,
        whole_file_digest: pass.whole_file_digest,
    })
}

fn preflight_parts<I>(declared_parts: u64, parts: I, budget: &Budget) -> Result<PartPass>
where
    I: IntoIterator<Item = ManifestStreamPart>,
{
    let mut count = 0u64;
    let mut logical_bytes = 0u64;
    let mut transcript = Sha256Writer::new();
    for part in parts {
        budget.check_time()?;
        count = count
            .checked_add(1)
            .ok_or_else(|| Error::new(ErrorCode::LimitCount))?;
        if count > declared_parts {
            return Err(Error::new(ErrorCode::LimitCount));
        }
        validate_part(&part, budget)?;
        logical_bytes = add_logical(logical_bytes, part.length, budget)?;
        transcript.update(&encode_part(&part));
    }
    if count != declared_parts {
        return Err(Error::new(ErrorCode::LimitCount));
    }
    Ok(PartPass {
        logical_bytes,
        part_transcript: transcript.finish(),
    })
}

fn verify_content<I, S>(
    declared_parts: u64,
    parts: I,
    source: &mut S,
    budget: &Budget,
) -> Result<ContentPass>
where
    I: IntoIterator<Item = ManifestStreamPart>,
    S: ManifestChunkSource + ?Sized,
{
    let mut count = 0u64;
    let mut logical_bytes = 0u64;
    let mut whole_file = Sha256Writer::new();
    let mut transcript = Sha256Writer::new();
    for part in parts {
        budget.check_time()?;
        let part_index = count;
        count = count
            .checked_add(1)
            .ok_or_else(|| Error::new(ErrorCode::LimitCount))?;
        if count > declared_parts {
            return Err(Error::new(ErrorCode::LimitCount));
        }
        validate_part(&part, budget)?;
        logical_bytes = add_logical(logical_bytes, part.length, budget)?;
        transcript.update(&encode_part(&part));

        let mut emitted = 0u64;
        let mut chunk_hash = ObjectHashWriter::new(
            ObjectKind::Chunk,
            MAX_CHUNK_BYTES as usize,
            MAX_METADATA_BYTES as usize,
        );
        let mut consume = |bytes: &[u8]| -> Result<()> {
            budget.check_time()?;
            if bytes.len() > budget.limits.max_memory_bytes {
                return Err(Error::new(ErrorCode::LimitMemory));
            }
            let length = u64::try_from(bytes.len())
                .map_err(|_| Error::new(ErrorCode::ManifestChunkLengthInvalid))?;
            emitted = emitted
                .checked_add(length)
                .ok_or_else(|| Error::new(ErrorCode::ManifestChunkLengthInvalid))?;
            if emitted > part.length {
                return Err(Error::new(ErrorCode::ManifestChunkLengthInvalid));
            }
            chunk_hash.update(bytes)?;
            whole_file.update(bytes);
            Ok(())
        };
        source.stream_chunk(part_index, &part, &mut consume)?;
        if emitted != part.length {
            return Err(Error::new(ErrorCode::ManifestChunkLengthInvalid));
        }
        if chunk_hash.finish()? != part.chunk {
            return Err(Error::new(ErrorCode::ObjectIdMismatch));
        }
    }
    if count != declared_parts {
        return Err(Error::new(ErrorCode::LimitCount));
    }
    Ok(ContentPass {
        logical_bytes,
        whole_file_digest: whole_file.finish(),
        part_transcript: transcript.finish(),
    })
}

fn validate_part(part: &ManifestStreamPart, budget: &Budget) -> Result<()> {
    if part.chunk.kind != ObjectKind::Chunk {
        return Err(Error::new(ErrorCode::ObjectReferenceKindMismatch)
            .with_stage(ValidationStage::KnownSchema));
    }
    if part.length == 0 {
        return Err(Error::new(ErrorCode::ManifestChunkLengthInvalid));
    }
    manifest_limit(
        "chunk-payload-bytes",
        part.length,
        budget.limits.max_chunk_bytes,
        ErrorCode::ManifestChunkLengthInvalid,
        2,
    )?;
    Ok(())
}

fn add_logical(current: u64, length: u64, budget: &Budget) -> Result<u64> {
    let sum = current
        .checked_add(length)
        .ok_or_else(|| Error::new(ErrorCode::LimitLogicalBytes))?;
    manifest_limit(
        "logical-file-bytes",
        sum,
        budget.limits.max_logical_bytes,
        ErrorCode::LimitLogicalBytes,
        2,
    )?;
    Ok(sum)
}

fn manifest_limit(
    name: &'static str,
    value: u64,
    configured: u64,
    code: ErrorCode,
    layer: u8,
) -> Result<()> {
    enforce_hard_limit_context(name, value, configured, code, layer).map(|_| ())
}

fn validate_chunk_profile(
    profile: &ProfileRef,
    registry: &Registry,
    operation: Operation,
) -> Result<()> {
    let entry = registry
        .profile(profile)
        .ok_or_else(|| Error::new(ErrorCode::ProfileUnknown))?;
    if entry.family != "chunking" {
        return Err(Error::new(ErrorCode::SchemaFieldInvalid));
    }
    registry.check_profile(profile, "chunking", operation)
}

fn validate_manifest_assignments(registry: &Registry, operation: Operation) -> Result<()> {
    for assignment in [
        RegistryAssignment::ObjectKind(ObjectKind::ContentManifest.code()),
        RegistryAssignment::ObjectKind(ObjectKind::Chunk.code()),
        RegistryAssignment::HashAlgorithm(1),
        RegistryAssignment::CommonField(0),
        RegistryAssignment::CommonField(1),
        RegistryAssignment::CommonField(2),
    ] {
        registry.check_assignment_if_present(assignment, operation)?;
    }
    for (rule, fields) in [
        ("content-manifest", &[16, 17, 18, 19][..]),
        ("typed-digest", &[0, 1][..]),
        ("profile-ref", &[0, 1, 2][..]),
        ("chunk-part", &[0, 1][..]),
        ("object-ref", &[0, 1, 2, 3][..]),
    ] {
        for &code in fields {
            registry.check_assignment_if_present(
                RegistryAssignment::KindField {
                    cddl_rule: rule,
                    code,
                },
                operation,
            )?;
        }
    }
    Ok(())
}

fn encode_part(part: &ManifestStreamPart) -> Vec<u8> {
    let mut output = Vec::with_capacity(56);
    push_head(&mut output, 5, 2);
    push_head(&mut output, 0, 0);
    push_object_ref(&mut output, part.chunk);
    push_head(&mut output, 0, 1);
    push_head(&mut output, 0, part.length);
    output
}

fn write_typed_digest<W: Write>(sink: &mut CanonicalSink<W>, digest: &[u8; 32]) -> Result<()> {
    sink.head(5, 2)?;
    sink.head(0, 0)?;
    sink.head(0, 1)?;
    sink.head(0, 1)?;
    sink.head(2, 32)?;
    sink.bytes(digest)
}

fn write_profile<W: Write>(sink: &mut CanonicalSink<W>, profile: &ProfileRef) -> Result<()> {
    sink.head(5, 3)?;
    sink.head(0, 0)?;
    sink.head(3, profile.namespace().len() as u64)?;
    sink.bytes(profile.namespace().as_bytes())?;
    sink.head(0, 1)?;
    sink.head(3, profile.id().len() as u64)?;
    sink.bytes(profile.id().as_bytes())?;
    sink.head(0, 2)?;
    sink.head(0, u64::from(profile.major()))
}

fn push_object_ref(output: &mut Vec<u8>, reference: ObjectRef) {
    push_head(output, 5, 4);
    push_head(output, 0, 0);
    push_head(output, 0, 1);
    push_head(output, 0, 1);
    push_head(output, 0, u64::from(reference.kind.code()));
    push_head(output, 0, 2);
    push_head(output, 0, 1);
    push_head(output, 0, 3);
    push_head(output, 2, 32);
    output.extend_from_slice(&reference.digest);
}

fn push_head(output: &mut Vec<u8>, major: u8, value: u64) {
    let (bytes, length) = cbor_head(major, value);
    output.extend_from_slice(&bytes[..length]);
}

fn cbor_head(major: u8, value: u64) -> ([u8; 9], usize) {
    let mut bytes = [0u8; 9];
    let prefix = major << 5;
    match value {
        0..=23 => {
            bytes[0] = prefix | value as u8;
            (bytes, 1)
        }
        24..=0xff => {
            bytes[0] = prefix | 24;
            bytes[1] = value as u8;
            (bytes, 2)
        }
        0x100..=0xffff => {
            bytes[0] = prefix | 25;
            bytes[1..3].copy_from_slice(&(value as u16).to_be_bytes());
            (bytes, 3)
        }
        0x1_0000..=0xffff_ffff => {
            bytes[0] = prefix | 26;
            bytes[1..5].copy_from_slice(&(value as u32).to_be_bytes());
            (bytes, 5)
        }
        _ => {
            bytes[0] = prefix | 27;
            bytes[1..9].copy_from_slice(&value.to_be_bytes());
            (bytes, 9)
        }
    }
}
